packages/vm/src/imports.rs

```diff
@@ -95,156 +82,169 @@ mod errors {
-    let key = match read_region(ctx, key_ptr, MAX_LENGTH_DB_KEY) {
-        Ok(data) => data,
-        Err(VmError::RegionLengthTooBig { .. }) => return errors::REGION_READ_LENGTH_TOO_BIG,
-        Err(_) => return errors::REGION_READ_UNKNOWN,
-    };
-    match with_storage_from_context::<S, Q, _, ()>(ctx, |store| {
-        store
-            .remove(&key)
-            .or_else(|_| Err(make_backend_err("Error removing database key from backend")))
-    }) {
-        Ok(_) => errors::NONE,
-        Err(VmError::UninitializedContextData { .. }) => errors::NO_CONTEXT_DATA,
-        Err(_) => errors::DB_UNKNOWN,
-    }
+pub fn do_remove<S: Storage, Q: Querier>(ctx: &mut Ctx, key_ptr: u32) -> VmResult<i32> {
+    let key = read_region!(ctx, key_ptr, MAX_LENGTH_DB_KEY);
+    with_storage_from_context::<S, Q, _, ()>(ctx, |store| Ok(store.remove(&key)?))
+        .and(Ok(errors::NONE))
 }
 
 pub fn do_canonicalize_address<A: Api>(
     api: A,
     ctx: &mut Ctx,
     human_ptr: u32,
     canonical_ptr: u32,
-) -> i32 {
-    let human_data = match read_region(ctx, human_ptr, MAX_LENGTH_HUMAN_ADDRESS) {
-        Ok(data) => data,
-        Err(VmError::RegionLengthTooBig { .. }) => return errors::REGION_READ_LENGTH_TOO_BIG,
-        Err(_) => return errors::REGION_READ_UNKNOWN,
-    };
+) -> VmResult<i32> {
+    let human_data = read_region!(ctx, human_ptr, MAX_LENGTH_HUMAN_ADDRESS);
     let human = match String::from_utf8(human_data) {
         Ok(human_str) => HumanAddr(human_str),
-        Err(_) => return errors::canonicalize::INVALID_INPUT,
+        Err(_) => return Ok(errors::canonicalize::INVALID_INPUT),
     };
-    match api.canonical_address(&human) {
-        Ok(canon) => match write_region(ctx, canonical_ptr, canon.as_slice()) {
-            Ok(()) => errors::NONE,
-            Err(VmError::RegionTooSmall { .. }) => errors::REGION_WRITE_TOO_SMALL,
-            Err(_) => errors::REGION_WRITE_UNKNOWN,
-        },
-        Err(_) => errors::canonicalize::UNKNOWN,
-    }
+    let canon = api.canonical_address(&human)?;
+    Ok(write_region!(ctx, canonical_ptr, canon.as_slice()))
 }
 
 pub fn do_humanize_address<A: Api>(
     api: A,
     ctx: &mut Ctx,
     canonical_ptr: u32,
     human_ptr: u32,
-) -> i32 {
-    let canonical = match read_region(ctx, canonical_ptr, MAX_LENGTH_CANONICAL_ADDRESS) {
-        Ok(data) => Binary(data),
-        Err(VmError::RegionLengthTooBig { .. }) => return errors::REGION_READ_LENGTH_TOO_BIG,
-        Err(_) => return errors::REGION_READ_UNKNOWN,
-    };
-    match api.human_address(&CanonicalAddr(canonical)) {
-        Ok(human) => match write_region(ctx, human_ptr, human.as_str().as_bytes()) {
-            Ok(()) => errors::NONE,
-            Err(VmError::RegionTooSmall { .. }) => errors::REGION_WRITE_TOO_SMALL,
-            Err(_) => errors::REGION_WRITE_UNKNOWN,
-        },
-        Err(_) => errors::humanize::UNKNOWN,
-    }
+) -> VmResult<i32> {
+    let canonical = Binary(read_region!(
```

Comment:
Huh, why do we need macros here?

Reply:
Because it's more convenient and readable. There's some control flow going on in there which can not be wrapped up into a function, and these code segments were just very bulky and repetitive. I think it's much more readable this way, because it's easier to see what the function is doing at a glance.